File: backend/app/api/v1/schemas/field_monitor.py
"""Pydantic schemas for Field Monitor API."""

from typing import Literal

from pydantic import BaseModel, Field


class FieldMonitorRequest(BaseModel):
    """Request to monitor a township's plots."""

    township_id: str = Field(min_length=1, max_length=64, pattern=r"^[a-z0-9_-]+$")
    season: Literal["monsoon", "dry"] = "monsoon"
    year: int = Field(default=2025, ge=2020, le=2030)


class PlotLocationResponse(BaseModel):
    """Geographic location of a plot."""

    latitude: float
    longitude: float


class PlotObservationResponse(BaseModel):
    """A single SAR observation for a plot."""

    date: str
    observed_vh_db: float
    expected_vh_db: float


class ComplianceInfoResponse(BaseModel):
    """Compliance scoring breakdown."""

    status: str
    score: float
    planting_detected: bool
    crop_match: bool
    phenology_match: float


class PlotAlertResponse(BaseModel):
    """An alert from a plot deviation."""

    alert_id: str
    plot_id: str
    farmer_name: str
    alert_type: str
    severity: str
    message: str
    created_date: str


class MonitoredPlotResponse(BaseModel):
    """A single monitored plot with full details."""

    plot_id: str
    farmer_name: str
    location: PlotLocationResponse
    area_ha: float
    recommended_crop: str
    compliance: ComplianceInfoResponse
    observations: list[PlotObservationResponse]
    alerts: list[PlotAlertResponse]


class FieldMonitorSummaryResponse(BaseModel):
    """Full monitoring summary for a township."""

    township_id: str
    season: str
    year: int
    total_plots: int
    compliant_count: int
    warning_count: int
    deviation_count: int
    compliance_rate: float
    total_area_ha: float
    active_alerts: int
    plots: list[MonitoredPlotResponse]
    alerts: list[PlotAlertResponse]

File: frontend/src/types/field-monitor.ts
/** Compliance status of a monitored plot. */
export type ComplianceStatus = "compliant" | "warning" | "deviation";

/** Alert severity level. */
export type AlertSeverity = "low" | "medium" | "high" | "critical";

/** Geographic location of a plot. */
export interface PlotLocation {
  latitude: number;
  longitude: number;
}

/** A single SAR observation for a plot. */
export interface PlotObservation {
  date: string;
  observed_vh_db: number;
  expected_vh_db: number;
}

/** Compliance scoring breakdown. */
export interface ComplianceInfo {
  status: ComplianceStatus;
  score: number;
  planting_detected: boolean;
  crop_match: boolean;
  phenology_match: number;
}

/** An alert from a plot deviation. */
export interface PlotAlert {
  alert_id: string;
  plot_id: string;
  farmer_name: string;
  alert_type: string;
  severity: AlertSeverity;
  message: string;
  created_date: string;
}

/** A single monitored plot with full details. */
export interface MonitoredPlot {
  plot_id: string;
  farmer_name: string;
  location: PlotLocation;
  area_ha: number;
  recommended_crop: string;
  compliance: ComplianceInfo;
  observations: PlotObservation[];
  alerts: PlotAlert[];
}

/** Full monitoring summary for a township. */
export interface FieldMonitorSummary {
  township_id: string;
  season: string;
  year: number;
  total_plots: number;
  compliant_count: number;
  warning_count: number;
  deviation_count: number;
  compliance_rate: number;
  total_area_ha: number;
  active_alerts: number;
  plots: MonitoredPlot[];
  alerts: PlotAlert[];
}

/** Request to monitor a township. */
export interface FieldMonitorRequest {
  township_id: string;
  season?: "monsoon" | "dry";
  year?: number;
}
